numpy/core/arrayprint.py

```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%' + signchar + '%d.%de' % (max_str_len, self.precision)
         else:
-            format = '%%.%df' % (self.precision,)
-            if len(non_zero):
-                precision = max([_digits(x, self.precision, format)
-                                 for x in non_zero])
+            if len(non_zero) and self.precision > 0:
+                precision = self.precision
+                trim_zero = lambda s: precision - (len(s) - len(s.rstrip('0')))
+                fmt = '%%.%df' % (precision,)
+                precision = max(trim_zero(fmt % x) for x in abs_non_zero)
             else:
                 precision = 0
-            precision = min(self.precision, precision)
-            self.max_str_len = len(str(int(max_val))) + precision + 2
+
+            int_len = len(str(int(max_val)))
+            signpos = self.sign != '-' or (len(str(int(min_val_sgn))) > int_len)
+            max_str_len = signpos + int_len + 1 + precision
+
             if _nc.any(special):
-                self.max_str_len = max(self.max_str_len,
-                                       len(_nan_str),
-                                       len(_inf_str)+1)
-            if self.sign:
-                format = '%#+'
-            else:
-                format = '%#'
-            format = format + '%d.%df' % (self.max_str_len, precision)
+                neginf = self.sign != '-' or _nc.any(data[hasinf] < 0)
+                nanlen = len(_format_options['nanstr'])
+                inflen = len(_format_options['infstr']) + neginf
+                max_str_len = max(max_str_len, nanlen, inflen)
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%#' + signchar + '%d.%df' % (max_str_len, precision)
 
-        self.special_fmt = '%%%ds' % (self.max_str_len,)
+        self.special_fmt = '%%%ds' % (max_str_len,)
         self.format = format
 
     def __call__(self, x, strip_zeros=True):
         from . import numeric as _nc
 
         with _nc.errstate(invalid='ignore'):
             if isnan(x):
-                if self.sign:
-                    return self.special_fmt % ('+' + _nan_str,)
+                nan_str = _format_options['nanstr']
+                if self.sign == '+':
+                    return self.special_fmt % ('+' + nan_str,)
                 else:
-                    return self.special_fmt % (_nan_str,)
+                    return self.special_fmt % (nan_str,)
             elif isinf(x):
+                inf_str = _format_options['infstr']
                 if x > 0:
-                    if self.sign:
-                        return self.special_fmt % ('+' + _inf_str,)
+                    if self.sign == '+':
+                        return self.special_fmt % ('+' + inf_str,)
                     else:
-                        return self.special_fmt % (_inf_str,)
+                        return self.special_fmt % (inf_str,)
```

Comment:
Should we not include a space for `sign==' '`? I.e., maybe
```
return self.special_fmt % (('' if self.sign == '-' else self.sign) + inf_str,)
```

Reply:
I think we are OK because of the way `special_fmt` is defined. If `sign=' '` then the `max_str_len` will be one longer, so we will get the necessary prefix space.